### PR TITLE
Impl `ops::Not` for `expr`

### PR DESCRIPTION
**Pull request by Jimexist**
# Which issue does this PR close?

Closes #.

 # Rationale for this change

Impl `ops::Not` for `expr` to make it more generic

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

**Comment by alamb** [comment]
This appears to also have code from https://github.com/apache/arrow-datafusion/pull/761 in it

**Review by alamb — APPROVED** [review]
I think this one looks good -- it seems to have gotten lost in the shuffle / list of PRs. Sorry about that @Jimexist

File: python/src/functions.rs [hunk 1]
```diff
@@ -20,7 +20,7 @@ use crate::udf;
 use crate::{expression, types::PyDataType};
 use datafusion::arrow::datatypes::DataType;
 use datafusion::logical_plan;
-use pyo3::{prelude::*, wrap_pyfunction};
+use pyo3::{prelude::*, types::PyTuple, wrap_pyfunction};
```
Comment (alamb) [inline python/src/functions.rs:23]: Are the changes to python intended for this PR @Jimexist ?

Comment (Jimexist) [reply python/src/functions.rs]: not really, let me rebase.

**Pull request merged**